### DISCUSSION TITLE
테스트 DB 를 따로 구축해야 할 것 같습니다

### DISCUSSION THREAD

**Discussion by Nine-JH** [petit-a-petit/MIML#18] [General]
# 문제발생
테스트 자동화를 구축 후 첫 PR을 날릴 때 테스트 실패가 발생했습니다. 이유는 DB 설정정보 누락이었습니다.
```text
Caused by: java.sql.SQLNonTransientConnectionException: Cannot load connection class because of underlying exception: 
com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${LOCAL_IP}:${MYSQL_LOCAL_PORT}'.
```
<br>
* 해당 문제는 github action token 에 해당 파라미터를 입력하면 해결 할 수 있습니다.

### 1.  Github Acttion 에서 DB 인스턴스를 생성하지 않는다면 Connection 문제가 발생할 수 있다.

<br>

### 2. 테스트 환경 격리
github action 을 제외하고도 발생하는 문제입니다. 현재 테스트를 위한 설정이 따로 존재하지 않습니다. 이로 인해 테스트의 DB 의존성은 로컬 DB를 향하게 됩니다.
![261227138-b2dc76f1-4f6b-4586-b7cc-24e36f4baa3d](https://github.com/petit-a-petit/MIML/assets/139187207/be64514e-d4b9-4fd2-8c81-27f7cb3dfebb)

위와 같은 환경은 격리성에 문제를 일으킵니다.

* 로컬 DB에 저장된 데이터에 따라 테스트 성공/실패 여부가 달라질 수 있음.
* 테스트 context에서 db에 영향을 주는 코드가 존재 시 LocalDB에 그대로 영향을 끼치게 됨.

<br>
<br>
<br>

# 해결 방안
### 장점 
* 외부에서 제어가 가능

### 단점
* 귀찮다.
<br>

### 장점
* 매우 간단하다

### 단점
* 외부 제어가 불가능하다.

<br>
<br>

추가적인 해결방안이 있으시면 알려주세요

**Comment by Jeonwoojae** [comment]
### github action 가상 서버에 MySQL 구축하기
- docker container를 명령어로 직접 띄우셔도 되고 있는 플러그인을 쓰셔도 될 것 같습니다!
- 단 해당 방법의 경우 말씀하셨던 로컬과 테스트 환경의 DB 데이터 차이로 인한 테스트 실패는 방지할 수 없을 것 같습니다(제가 이해했던 내용이 맞나요..?😢)

**Reply by Nine-JH** [reply]
네 맞아요 ㅋㅋ 플러그인이나 도커를 쓰는건 좋은 방법인 것 같아요!

**Comment by minju412** [comment]
제가 진행했던 이전 프로젝트를 확인해보니 `application.yml` 안에서 profile 설정을 하고 `build.gradle`에서 테스트 환경을 명시해줬던 것 같아요!
이 방법이 위에 적힌 2번 방법인 것 같은데, 단점으로 적어주신 "외부 제어"라는게 어떤 상황에서 필요한지 감이 잘 오지 않아 여쭤보고 싶습니다..!

**Reply by Nine-JH** [reply]
조금 추상적으로 들렸을 수 있네요! 간단하게 말하자면 DB 설정이나, 데이터 로깅 등을 말하는것이었어요
